### PR TITLE
KEP-466: Address requests by their hash

### PR DESCRIPTION
**Pull request by isabelsavannah**

### CODE REVIEW

File: pbft/pbft_operation.cpp [hunk 1]
```diff
@@ -94,7 +137,7 @@ pbft_operation::end_commit_phase()
 operation_key_t
 pbft_operation::get_operation_key()
 {
-    return std::tuple<uint64_t, uint64_t, hash_t>(this->view, this->sequence, request_hash(this->request));
+    return std::tuple<uint64_t, uint64_t, hash_t>(this->view, this->sequence, this->request_hash);
```
Comment (ebruck) [inline pbft/pbft_operation.cpp]: you can reduce this to just:

    return {this->view, this->sequence, request_hash(this->request, this->request_hash)};

File: pbft/pbft_operation.hpp [hunk 2]
```diff
@@ -40,12 +39,10 @@ namespace bzn
     {
     public:
 
-        pbft_operation(uint64_t view, uint64_t sequence, pbft_request msg, std::shared_ptr<const std::vector<peer_address_t>> peers);
+        pbft_operation(uint64_t view, uint64_t sequence, bzn::hash_t request_hash, std::shared_ptr<const std::vector<peer_address_t>> peers);
```
Comment (ebruck) [inline pbft/pbft_operation.hpp]: const bzn::hash_t&  perhaps?

File: pbft/test/pbft_join_leave_test.cpp [hunk 3]
```diff
@@ -64,14 +90,17 @@ namespace bzn
...
+            preprepare.set_request_hash(expect_hash);
 
             // receiving node should send out prepare messsage to everyone
             for (auto const &p : TEST_PEER_LIST)
             {
                 EXPECT_CALL(*(mock_node),
                     send_message_str(bzn::make_endpoint(p),
-                        message_is_correct_type(PBFT_MSG_PREPARE, PBFT_REQ_NEW_CONFIG)))
+                        AllOf(message_has_correct_req_hash(expect_hash), message_has_correct_pbft_type(PBFT_MSG_PREPARE))))
                     .Times(Exactly(1));
```
Comment (paularchard) [inline pbft/test/pbft_join_leave_test.cpp]: and (message_has_req_with_correct_type(PBFT_REQ_NEW_CONFIG) ?

Comment (paularchard) [reply pbft/test/pbft_join_leave_test.cpp]: nevermind, I remembered that the prepare doesn't carry the request any more.

File: pbft/pbft.cpp [hunk 4]
```diff
@@ -74,16 +76,16 @@ pbft::start()
 
                 this->service->register_execute_handler(
                         [weak_this = this->weak_from_this(), fd = this->failure_detector]
-                                (const pbft_request& req, uint64_t sequence)
+                                (std::shared_ptr<pbft_operation> op)
                                         {
-                                            fd->request_executed(req);
+                                            fd->request_executed(op->request_hash);
```
[the remark this answers is not in the document]
Comment (isabelsavannah) [reply pbft/pbft.cpp]: Ach, that's true. Added a check to make that at least not crash for now.

**Pull request merged**